pkg/csrutil/csrutil.go

```diff
@@ -0,0 +1,245 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package csrutil contains types for CSR approval workflows.
+package csrutil
+
+import (
+	"crypto/x509"
+	"fmt"
+	"reflect"
+	"strings"
+
+	authorization "k8s.io/api/authorization/v1beta1"
+	capi "k8s.io/api/certificates/v1beta1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/cloud-provider-gcp/pkg/csrmetrics"
+	"k8s.io/klog"
+	certutil "k8s.io/kubernetes/pkg/apis/certificates/v1beta1"
+	"k8s.io/kubernetes/pkg/controller/certificates"
+)
+
+type RecognizeFunc func(csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool
+type ValidateFunc func(csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) (bool, error)
+type PreApproveFunc func(csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest, label string) error
+
+var nodeClientKeyUsages = []capi.KeyUsage{
+	capi.UsageKeyEncipherment,
+	capi.UsageDigitalSignature,
+	capi.UsageClientAuth,
+}
+
+var nodeServerKeyUsages = []capi.KeyUsage{
+	capi.UsageKeyEncipherment,
+	capi.UsageDigitalSignature,
+	capi.UsageServerAuth,
+}
+
+type ValidatorInterface interface {
```

Comment:
does this have to be an interface?
it seems to overlap with the `*Func` types above too.
could it be just a struct with Options and func fields?

Reply:
I should remove the bare func definitions - its an interface so that the HandleCSR() code does not have to know the internals of each validators context. such fields (eg. connection to gce endpoints) can be hidden in the struct that implements ValidatorInterface.